### PR TITLE
Update the builder plugin to make version flag optional and add new 'build-plugin-local' make target 

### PR DESCRIPTION
**Pull request by chandrareddyp**
### What this PR does / why we need it

1. Updates the command `tanzu builder cli compile --match <package_name> --version` logic to make the flag "--version" as optional and the help text of the flag.
2. New make targets `build-plugin-local` and `install-plugin-local` has added to give flexibility to users to build and install the given plugin locally.

### Which issue(s) this PR fixes

Fixes #https://github.com/vmware-tanzu/tanzu-framework/issues/899

### Describe testing done for PR

1. The command ` tanzu builder cli compile --match package --version $(make version) --ldflags "-X 'github.com/vmware-tanzu/tanzu-framework/pkg/v1/buildinfo.Version=v0.27.0-dev'"` tested with --ldflags having Version info, works fine.
2. The command ` tanzu builder cli compile --match package ---ldflags "-X 'github.com/vmware-tanzu/tanzu-framework/pkg/v1/buildinfo.Version=v0.27.0-dev'"` tested without --version flag, works fine.
3. The new make target `build-plugin-local` tested: With valid plugin name `make build-plugin-local PLUGIN_NAME=package` works fine, builds the given plugin. Also tested with empty and in-valid PLUGIN_NAME, it does throw an error saying: `The PLUGIN_NAME: '' is not valid or not exists or empty, please provide valid PLUGIN_NAME.`
4. The new make target `install-plugin-local` tested: With valid plugin name `make install-plugin-local PLUGIN_NAME=login` works fine, build and install the given plugin. Also tested with empty and in-valid PLUGIN_NAME, it does throw an error saying: `The PLUGIN_NAME: '' is not valid or not exists or empty, please provide valid PLUGIN_NAME.`

### Release note

```release-note
The command `tanzu builder cli compile` has been updated to make the flag `--version` as optional
New make target  `build-plugin-local` has been added to build the given plugin locally
New make target `install-plugin-local` has been added to build and install the given plugin locally
```

### CODE REVIEW

**Review by marckhouzam — COMMENTED** [review]
This works for me.  One comment in-line.

File: pkg/v1/builder/command/cli_compile.go [hunk 1]
```diff
@@ -239,6 +239,7 @@ func buildPlugin(path string, arch cli.Arch, id string) (plugin, error) {
 
 	var modPath string
 
+	ldflags += " -X 'github.com/vmware-tanzu/tanzu-framework/pkg/v1/buildinfo.Version=" + version + "' "
```
[the remark this answers is not in the document]
Comment (chandrareddyp) [reply pkg/v1/builder/command/cli_compile.go:242]: That's a good catch; we will definitely use https://github.com/vmware-tanzu/tanzu-framework/blob/main/cli/runtime/buildinfo/buildinfo.go after the refactoring, for now, I will add a comment and also add the Version from  https://github.com/vmware-tanzu/tanzu-framework/blob/main/cli/runtime/buildinfo/buildinfo.go, so that we can avoid bug if in case refactoring ignores adding Version from cli/runtime/buildinfo/buildinfo.go

[the remark this answers is not in the document]
Comment (anujc25) [reply pkg/v1/builder/command/cli_compile.go:242]: I am not sure we should do this hardcoded string here because it requires change in builder plugin everytime this `buildinfo` gets moved to different location.

Also, this is just general recommendation that plugins use to define their versions but plugin developer can hardcode the plugin version or rely on some other means to define plugin version in the PluginDescriptor. So, assuming this might not be good idea. 

We need to think about passing this type of version is even possible to build a specific plugin or not.

[the remark this answers is not in the document]
Comment (saji-pivotal) [reply pkg/v1/builder/command/cli_compile.go:242]: @chandrareddyp I don't think we should be hardcoding an ldflag in the code. IMO we should be able to set `buildinfo.Version` by doing something like what's being done in the Makefile - https://github.com/vmware-tanzu/tanzu-framework/blob/main/Makefile#L87-L115

[the remark this answers is not in the document]
Comment (marckhouzam) [reply pkg/v1/builder/command/cli_compile.go:242]: > @chandrareddyp I don't think we should be hardcoding an ldflag in the code. IMO we should be able to set `buildinfo.Version` by doing something like what's being done in the Makefile 

The problem with that is that if you just call `tanzu builder cli compile --match package --version v0.1.2` the `--version` flag will be ignored, which seems wrong from a user's point-of-view.  This is related to why @chandrareddyp was asking what this `--version` flag is really meant to do.

> I am not sure we should do this hardcoded string here because it requires change in builder plugin everytime this `buildinfo` gets moved to different location.

True.

> Also, this is just general recommendation that plugins use to define their versions but plugin developer can hardcode the plugin version or rely on some other means to define plugin version in the PluginDescriptor. So, assuming this might not be good idea.
> We need to think about passing this type of version is even possible to build a specific plugin or not.

Yeah, that is tricky.  It comes back to "what is the `--version` flag supposed to do?"

[the remark this answers is not in the document]
Comment (anujc25) [reply pkg/v1/builder/command/cli_compile.go:242]: > The problem with that is that if you just call tanzu builder cli compile --match package --version v0.1.2 the --version flag will be ignored, which seems wrong from a user's point-of-view.

I was not involved in the development of this plugin/command, however, Based on the help-text for version, it seems it is meant only for core-cli version and not for plugin version. ` --version string       Version of the root cli (required)`

I think, the main issue here is that `--version` flag should not be a required flag as we expect user to just build the plugin without core-cli.

[the remark this answers is not in the document]
Comment (chandrareddyp) [reply pkg/v1/builder/command/cli_compile.go]: The  `--version` flag is not required to compile the plugin, so we have made changes to make `--version` flag as optional.

**Review by anujc25 — COMMENTED** [review]
`build-plugin-local` will compile the plugin and put it in the `artifacts` directory within the repository.

We should add `install-plugin-local` that installs the plugin as well. Because as a end user I would like to do some change in the plugin, then I want to build as well as install the plugin to test my change.

**Review by marckhouzam — COMMENTED** [review]
Nit: 
* git recommends that the commit title (the first line) be 50 characters or less.  This is useful because that line is used in many different git outputs (see the first paragraph from https://git-scm.com/docs/git-commit#_discussion)
* as for the text of the commit, it is recommended that it wraps at 72 characters.  If not, it makes `git log` hard to read, among other git output.

File: Makefile [hunk 2]
```diff
@@ -229,6 +230,10 @@ build-plugin-admin-with-oci-discovery: ${CLI_ADMIN_JOBS_OCI_DISCOVERY} publish-a
 .PHONY: build-plugin-admin-with-local-discovery
 build-plugin-admin-with-local-discovery: ${CLI_ADMIN_JOBS_LOCAL_DISCOVERY} publish-admin-plugins-all-local ## Build Tanzu CLI admin plugins with Local standalone discovery
 
+.PHONY: build-plugin-local
+build-plugin-local: ## Build plugin's locally for given plugin name (match-expression) and path
+	tanzu builder cli compile --match "$(PLUGIN_NAME)" --version $(BUILD_VERSION) --ldflags "$(LD_FLAGS)" --path $(PLUGIN_PATH) --target local --artifacts artifacts/${GOHOSTOS}/${GOHOSTARCH}/cli
```
Comment (marckhouzam) [inline Makefile:235]: 1. I don't think we should use `tanzu builder` because the `builder` plugin may not be built, or up to date.
Elsewhere we use `$(GO) run ./cmd/cli/plugin-admin/builder` which will compile and run the `builder` plugin on the spot. Note that the refactoring of the `builder` changes that a little but I can adapt my PR on top of yours: #3349
 
1. I'm a little worried that it is not clear that the user needs to set PLUGIN_NAME.  Without it nothing gets built but it give the impression that it may have built something.  Maybe we need an explicit logic check to handle this case?  I'm not sure.

File: pkg/v1/builder/template/plugintemplates/Makefile.tmpl [hunk 3]
```diff
@@ -11,6 +11,8 @@ GOARCH ?= $(shell go env GOARCH)
 GOHOSTOS ?= $(shell go env GOHOSTOS)
 GOHOSTARCH ?= $(shell go env GOHOSTARCH)
 
+PLUGIN_PATH ?= ./cmd/cli/plugin
```
Comment (marckhouzam) [inline pkg/v1/builder/template/plugintemplates/Makefile.tmpl:14]: I think this should be `./cmd/plugin` for the generated plugins

File: pkg/v1/builder/template/plugintemplates/Makefile.tmpl [hunk 4]
```diff
@@ -39,7 +41,11 @@ go.mod go.sum: $(GO_SRCS)
 
 .PHONY: build-local
 build-local: ## Build the plugin
-	tanzu builder cli compile --version $(BUILD_VERSION) --ldflags "$(LD_FLAGS)" --path ./cmd/plugin --target local --artifacts artifacts/${GOHOSTOS}/${GOHOSTARCH}/cli
+	tanzu builder cli compile --version $(BUILD_VERSION) --ldflags "$(LD_FLAGS)" --path $(PLUGIN_PATH) --target local --artifacts artifacts/${GOHOSTOS}/${GOHOSTARCH}/cli
+
+.PHONY: build-plugin-local
+build-plugin-local: ## Build plugin's locally for given plugin name (match-expression) and path
+	tanzu builder cli compile --match "$(PLUGIN_NAME)" --version $(BUILD_VERSION) --ldflags "$(LD_FLAGS)" --path $(PLUGIN_PATH) --target local --artifacts artifacts/${GOHOSTOS}/${GOHOSTARCH}/cli
```
Comment (marckhouzam) [inline pkg/v1/builder/template/plugintemplates/Makefile.tmpl:48]: Why do we want this target for the generated plugins?  It looks the same as `build-local` except for the `--match` flag.  However, I think there is usually only a single plugin in these generated repos, so using `--match` may not be useful.  But I'm just learning about this stuff so I can very well be wrong.

**Review by marckhouzam — CHANGES_REQUESTED** [review]
A couple of things to clarify

File: pkg/v1/builder/template/plugintemplates/Makefile.tmpl (hunk 3 above)
Comment (chandrareddyp) [reply pkg/v1/builder/template/plugintemplates/Makefile.tmpl:14]: PLUGIN_PATH represents where it needs to search for plugin files.

File: Makefile (hunk 2 above)
Comment (chandrareddyp) [reply Makefile:235]: > I'm a little worried that it is not clear that the user needs to set PLUGIN_NAME. Without it nothing gets built but it give the impression that it may have built something. Maybe we need an explicit logic check to handle this case? I'm not sure.
I don't think there is a way to add help text for the make target, as it's being used by developers they must be aware of usage.

Comment (chandrareddyp) [reply Makefile:235]: > I don't think we should use tanzu builder because the builder plugin may not be built, or up to date.
> Elsewhere we use $(GO) run ./cmd/cli/plugin-admin/builder which will compile and run the builder plugin on the spot. Note that the refactoring of the builder changes that a little but I can adapt my PR on top of yours: https://github.com/vmware-tanzu/tanzu-framework/pull/3349

@anujc25 any comments on this?

File: pkg/v1/builder/template/plugintemplates/Makefile.tmpl (hunk 3 above)
Comment (chandrareddyp) [reply pkg/v1/builder/template/plugintemplates/Makefile.tmpl:14]: After our discuss, i agree that it should be ./cmd/plugin

Comment (chandrareddyp) [reply pkg/v1/builder/template/plugintemplates/Makefile.tmpl]: changes done.

File: pkg/v1/builder/template/plugintemplates/Makefile.tmpl (hunk 4 above)
Comment (chandrareddyp) [reply pkg/v1/builder/template/plugintemplates/Makefile.tmpl]: After discussing with Anuj, we decided not to add new make targets in Makefile.tmpl

File: Makefile (hunk 2 above)
Comment (chandrareddyp) [reply Makefile]: We have added the validation to validate the PLUGIN_NAME, it builds and install only given plugin, if plugin name is empty then it throws an error.

**Comment by chandrareddyp** [comment]
> `build-plugin-local` will compile the plugin and put it in the `artifacts` directory within the repository.
> 
> We should add `install-plugin-local` that installs the plugin as well. Because as a end user I would like to do some change in the plugin, then I want to build as well as install the plugin to test my change.

Implementation has updated to add new make target `install-plugin-local` to build and update given plugin locally.

File: Makefile [hunk 5]
```diff
@@ -240,6 +241,20 @@ build-plugin-admin-with-oci-discovery: ${CLI_ADMIN_JOBS_OCI_DISCOVERY} publish-a
 .PHONY: build-plugin-admin-with-local-discovery
 build-plugin-admin-with-local-discovery: ${CLI_ADMIN_JOBS_LOCAL_DISCOVERY} publish-admin-plugins-all-local ## Build Tanzu CLI admin plugins with Local standalone discovery
 
+.PHONY: build-plugin-local
+build-plugin-local: ## Build given CLI Plugin locally, needs PLUGIN_NAME
```
Comment (anujc25) [inline Makefile:245]: This target needs to be linked with `prepare-builder`. I am seeing below error otherwise

```
$ make build-plugin-local PLUGIN_NAME=login
bash: /Users/anujc/Documents/code/tanzu-framework/bin/builder: No such file or directory
make: *** [build-plugin-local] Error 127
```

**Review by anujc25 — COMMENTED** [review]
One comment. But otherwise looks good.

File: Makefile (hunk 5 above)
Comment (chandrareddyp) [reply Makefile]: for some reason, i don't see the error even in the new terminal, anyhow I have updated the target with prepare-builder

**Review by marckhouzam — APPROVED** [review]
LGTM

Thanks!

File: Makefile (hunk 5 above)
Comment (marckhouzam) [reply Makefile]: The builder used by the Makefile is in `./bin/builder`.  If you have built it anytime by another make target, it will stay there.  This is probably why Chandra didn't see the error.

**Review by anujc25 — APPROVED** [review]
Thanks for the change. LGTM
